### PR TITLE
Database loader servlet

### PR DESCRIPTION
**Pull request by anamariamilcu**
The servlet loads the Books from the CSV file. I also added the python script that generated the CSV file that I used.
The books can be loaded just by simply going to /load-database in the web site or run the cron job called "Load books stocks"
at https://pantheon.corp.google.com/appengine/cronjobs?project=alphabooks-step-2020. This was configured using a cron.yaml file and running the command:
$ gcloud app deploy cron.yaml.

### CODE REVIEW

File: src/utils/databaseCreator.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  # Since I observed that Google Books API behaves quite weird and offers weird answers, I decided to add
+  # in our database the first 5 answers the API offers for each popular book that we had in our dataset,
+  # so when a user searches something, he could actually find books that are actually added in the databse 
+  # and can be ordered.
+  writer.writerow(["goodreads_id", "title", "author", "googlebooks_id1", "googlebooks_id2", "googlebooks_id3", "googlebooks_id4", "googlebooks_id5"])
+  for i in range(lenIds):
+    # goodreads accepts one request per second
+    time.sleep(1)
+    id1 = goodreadsId[i]
+    response = requests.get('https://www.goodreads.com/book/show/' + id1 + '.xml?key=c5WeMxjtRTJEknOl5E21g')
```
Comment (ak47na) [inline src/utils/databaseCreator.py]: Can you add a comment describing the response format for goodreads API?

File: src/utils/databaseCreator.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import csv
```
Comment (alexcepoi) [inline src/utils/databaseCreator.py]: As discussed, let's not commit this one-off script.

File: src/utils/databaseCreator.py (hunk 1 above)
Comment (anamariamilcu) [reply src/utils/databaseCreator.py]: Added more comments

**Review by ak47na — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  /**
+   * For each Google Books Api ID from the CSV file, it will create a Book entity
+   * and will add it to Datastore.
+   * @param HttpServletResponse{response}
+   * @throws IOException
+   */
+
+  private void loadBookStocksFromCSV(HttpServletResponse response) throws IOException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: Please extract this logic to a separate class and test it.

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+/**
+ * When user goes to /load-database, he needs to wait for the database to get loaded.
+ * Will parse the CSV file to create fictive book stocks.
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java:20]: What does it do if stocks already exist for the imported books?

Does it just skip them. Does it add new stocks?

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  private final Integer TOTAL_STOCK = 185;
+  private final Integer NO_OF_LIBRARIES = 37;
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  /**
+   * Given an ID, it creates a book Entity for the ID and adds it to Datastore.
+   * @param String{bookId}
+   */
+
+  private void addBookStockToDatastore(String bookId) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: Please extract this into a class and test it.

**Review by alexcepoi — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 4 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java:20]: It skips them. I was thinking that the cron job can be run anytime the library wants to add new books and they can just edit the CSV file.

File: src/utils/databaseCreator.py (hunk 2 above)
Comment (anamariamilcu) [reply src/utils/databaseCreator.py]: Done

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 3 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: Done

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 5 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: Done

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 4 above)
Comment (alexcepoi) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java:20]: Sure, just document the behavior in the comment here. The reader shouldn't have to read the implementation to understand how this should work.

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    File csvFile;
+    try {
+      csvFile = new File(this.getClass().getResource("/availableBooks.csv").toURI());
+    } catch (URISyntaxException e) {
+      System.out.println("File not found!");
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: nit: this seems redundant. given the error and the stacktrace.

File: src/test/java/com/google/sps/data/DatabaseHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  @Test
+  public void testLoadingDuplicateDataFromCSV() throws URISyntaxException, IOException {
+    File csvFile = new File(this.getClass().getResource("/examples.csv").toURI());
+    // Load CSV twice
+    databaseHandler.loadBookStocksFromCSV(csvFile);
+    databaseHandler.loadBookStocksFromCSV(csvFile);
+    Query query = new Query("Book").setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, "gqX7rQEACAAJ"));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    // Check if book was added just once to datastore.
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/DatabaseHandlerTest.java:93]: can you also check what's the stock?

File: src/test/java/com/google/sps/data/DatabaseHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    Assert.assertEquals(1, datastore.prepare(query).asList(FetchOptions.Builder.withDefaults()).size());
+  }
+
+    
+  @After
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/DatabaseHandlerTest.java]: nit: generally you'd put this method right affer setup()

This actually runs after each individual test, so it's presence at the bottom of the file isn't really more intuitive.

**Review by alexcepoi — APPROVED** [review]
Nice!

File: src/test/java/com/google/sps/data/DatabaseHandlerTest.java (hunk 7 above)
Comment (anamariamilcu) [reply src/test/java/com/google/sps/data/DatabaseHandlerTest.java:93]: I'm not sure what do you mean by this? Then do what with the stock?

File: src/test/java/com/google/sps/data/DatabaseHandlerTest.java (hunk 8 above)
Comment (anamariamilcu) [reply src/test/java/com/google/sps/data/DatabaseHandlerTest.java]: Fixed

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 6 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java]: Fixed

File: src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java (hunk 4 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/servlets/DatabaseLoaderServlet.java:20]: Fixed

**Pull request merged**